src/Connection/Resource/Stream.php

```diff
@@ -207,7 +207,27 @@ public function write(string $string): int
 
         $result = fwrite($this->stream, $string);
 
-        if ($result === false) {
+        if ($result === false || $result === 0) {
+            $metadata = $this->getMetadata();
+
+            if ($this->eof()) {
+                throw new RuntimeException('Connection closed by peer during write', 1);
+            }
+
+            if (!is_resource($this->stream)) {
+                throw new RuntimeException(
+                    'Stream resource is no longer valid',
+                    1
+                );
+            }
+
+            if ($metadata['timed_out']) {
+                throw new RuntimeException(
+                    'Stream has been timed out',
+                    2
+                );
+            }
+
             throw new RuntimeException('Unable to write to stream', 1);
```

Comment:
`fwrite()` returning `0` is valid state if EOF and timeout doesn't appear. This could be quite common on slower non-blocking connection especially when SSL is involved. I would suggest to return `0` and add small sleep to avoid busy loop #1596 #1613